src/ser.rs

```diff
@@ -0,0 +1,86 @@
+use crate::*;
+use serde::{
+    de::{Deserializer, Error, MapAccess, Visitor},
+    ser::{SerializeMap, Serializer},
+    Deserialize, Serialize,
+};
+use std::{fmt, marker::PhantomData};
+
+///////////////////////////////////////////// ~ser~~~ /////////////////////////////////////////////
+
+#[derive(Serialize)]
+#[serde(untagged)]
+#[serde(bound(serialize = "T::Kind: Serialize"))]
+enum NodeContentSer<'a, T: Types> {
+    Leaf(&'a str),
+    Branch(&'a [GreenNode<T>]),
+}
+
+impl<T: Types> Serialize for GreenNode<T>
+where
+    T::Kind: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(1))?;
+        if let Some(text) = self.leaf_text() {
+            map.serialize_entry(&self.kind(), &NodeContentSer::<T>::Leaf(text.as_str()))?;
```

Comment:
```suggestion
            map.serialize_entry(&self.kind(), text.as_str())?;
```

Reply:
I assume these are equivalent then? I originally used the `#[serde(untagged)]` enum for obvious-correctness.

Reply:
Yes, untagged newtype variants behave like the underlying type i.e. the NodeContentSer derive expands to:

```rust
impl<'a, T: Types> Serialize for NodeContentSer<'a, T>
where
    T::Kind: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match self {
            NodeContentSer::Leaf(leaf) => Serialize::serialize(leaf, serializer),
            NodeContentSer::Branch(branch) => Serialize::serialize(branch, serializer),
        }
    }
}
```